src/index.ts

```diff
@@ -0,0 +1,229 @@
+ *     are excluded unconditionally from the tutorial.
+ * 
+ * Example code:
+ * 
+ *     import { ExampleParser } from "./example_parser"
+ *     let text: string = "// BEGIN FRAGMENT: test\nHello.\n// END FRAGMENT"
+ *     let parser_inst: ExampleParser = new ExampleParser(text)
+ *     console.log(parser_inst.mapLines())
+ * 
+ */
+
+// A custom error type to be used with broken syntax
+export class ExampleSyntaxError extends Error {
+    constructor(message:string = '') {
+      super(message)
+      this.message = message
+    }
+}
+
+// Obtain a range of numbers as an array, including the end
+let range = function(start: number, end: number): Array<number> {
+    let src: Array<number> = Array.from(Array(end - start + 1).keys())
+    return src.map(x => x + start)
+}
+
+// Check if the string is empty or contains
+// only the space and tabulation characters
+function emptySpace(str: string) {
+    return str.trim().length === 0
+}
+
+// Remove empty strings at both start and end
+// in an array of lines
+export function trimEmptyLines(lines: Array<string>) {
+    // Remove the first and the last string if those are
+    // filled with spaces or empty
+    while (emptySpace(lines.at(0) || '')) lines.shift()
+    while (emptySpace(lines.at(-1) || '')) lines.pop()
+    return lines
+}
+
+// Clear line indentation for an array of strings containing the lines
+let clearIndentation = function(lines: Array<string>) {
+    // Calculate a padding common for all the lines
+    let commonIndent: number = getCommonStartPadding(lines)
+    // Return the lines with a common indentation removed
+    return lines.map(
+        line => {return line.slice(commonIndent) }
+    )
+}
+
+// Checks if whitespaces and tabs of a given prefix length are equal for all provided lines
+function spaceSectionEquals(lines: Array<string>, prefixLen: number) {
+    let prefixes = lines.map(line => {return line.slice(0, prefixLen)})
+    let prefixIsTrimmable = lines[0].slice(0, prefixLen).trim().length === 0
+    let prefixesAreEqual = prefixes.every( v => (v === lines[0].slice(0, prefixLen)) )
+    return prefixIsTrimmable && prefixesAreEqual
+}
+
+// Finds a minimum padding length
+function getCommonStartPadding(lines: Array<string>) {
+    let knownMaxLen = 0
+    let currentLen = 0
+    if (lines.length) {
+        while (true) {
+            currentLen++
+            if (spaceSectionEquals(lines, currentLen)) {
+                knownMaxLen = currentLen
+            } else break
+        }
+    }
+    return knownMaxLen
+}
+
+// Regular expressions, used to parse the examples
+const exampleRegExps: { [id: string] : RegExp; } = {
+    beginFragment: new RegExp('[\s\t]*(\/\/|#) (BEGIN FRAGMENT): ([A-Za-z_\-]+)[\s\t]*'),
+    endFragment: new RegExp('[\s\t]*(\/\/|#) (END FRAGMENT)[\s\t]*'),
+    beginEscape: new RegExp('[\s\t]*(\/\/|#) (BEGIN ESCAPE)[\s\t]*'),
+    endEscape: new RegExp('[\s\t]*(\/\/|#) (END ESCAPE)[\s\t]*'),
+}
```

Comment:
```suggestion
const exampleRegExps = {
    beginFragment: new RegExp('[\s\t]*(\/\/|#) (BEGIN FRAGMENT): ([A-Za-z_\-]+)[\s\t]*'),
    endFragment: new RegExp('[\s\t]*(\/\/|#) (END FRAGMENT)[\s\t]*'),
    beginEscape: new RegExp('[\s\t]*(\/\/|#) (BEGIN ESCAPE)[\s\t]*'),
    endEscape: new RegExp('[\s\t]*(\/\/|#) (END ESCAPE)[\s\t]*'),
} as const
```

That would be much type-stronger.

Reply:
Thanks!